src/main/logutils.ts

```diff
@@ -373,32 +376,132 @@ const determineArenaMatchResult = (line: LogLine): any[] => {
-}
+    const zoneName = line.args[2];
+    const mapId = parseInt(line.args[3], 10);
+
+    if (!(mapId in dungeonTimersByMapId && mapId in dungeonsByMapId)) {
+        console.error(`[ChallengeMode] Invalid/unsupported mapId for Challenge Mode dungeon: ${mapId} ('${zoneName}')`)
+    }
+
+    const dungeonAffixes = line.args[5].map((v: string) => parseInt(v, 10));
+
+    activeChallengeMode = new ChallengeModeDungeon(
+        videoStartDate.getTime(),    // Start time
+        dungeonTimersByMapId[mapId], // Dungeon timers
+        parseInt(line.args[2], 10),  // zoneId
+        mapId,                       // mapId
+        parseInt(line.args[4], 10),  // Keystone Level
+        dungeonAffixes,              // Array of affixes, as numbers
+    )
+
+    activeChallengeMode.addVideoSegment(new ChallengeModeVideoSegment(
+        VideoSegmentType.Trash, videoStartDate, 0
+    ));
+
+    console.debug("[ChallengeMode] Starting Challenge Mode instance")
+
+    metadata = {
+        name: line.args[1],
+        encounterID: parseInt(line.args[1], 10),
+        category: VideoCategory.MythicPlus,
+        zoneID: parseInt(line.args[5]),
+        duration: 0,
+        result: false,
+        challengeMode: activeChallengeMode
+    };
+
+    recorder.start();
+};
 
 /**
- * Handle a line from the WoW log. 
+ * Handle a log line for CHALLENGE_MODE_END
  */
-function handleEncounterStartLine (line: LogLine): void {
-    if (isChallengeModeActive) {
-        console.log("[Logutils] ENCOUNTER_START in an active Mythic Keystone dungeon is ignored.")
+function handleChallengeModeEndLine (line: LogLine): void {
+    if (!recorder.isRecording || !activeChallengeMode) {
         return;
     }
-    
+
+    if (playerCombatant) {
+        metadata.playerName = playerCombatant.name;
+        metadata.playerRealm = playerCombatant.realm;
+        metadata.playerSpecID = playerCombatant.specID;
+    }
+
+    // Add a few seconds so we reliably see the aftermath of a kill.
+    const overrun = 5;
+
+    const videoStopDate = line.date();
+    const milliSeconds = (videoStopDate.getTime() - videoStartDate.getTime());
+    const duration = Math.round(milliSeconds / 1000);
+
+    metadata.duration = duration + overrun;
+    metadata.result = Boolean(parseInt(line.args[1]));
+
+    combatantMap.clear();
     playerCombatant = undefined;
-    const encounterID = parseInt(line.args[1], 10);
+
+    // The actual log duration of the dungeon, from which keystone upgrade
+    // levels can be calculated
+    activeChallengeMode.duration = Math.round(parseInt(line.args[4], 10) / 1000);
+
+    // Calculate whether the key was timed or not
+    activeChallengeMode.timed = calculateKeystoneCompletionResult(activeChallengeMode.allottedTime, activeChallengeMode.duration) > 0;
+
+    console.debug("[ChallengeMode] Ending current video segment")
+    activeChallengeMode.endVideoSegment(videoStopDate);
+
+    // If last video segment is less than 10 seconds long, discard it.
```

Comment:
Perhaps I could rename `VideoSegment` to `TimelineSegment`? I agree it could be confusing.

Reply:
I think I prefer videoSegment, but not overly fussed - would be happy with a comment here

Reply:
I renamed the whole video segment thing because I did find it confusing myself but did't have a better alternative at the time. It's now called timeline segments. 🤷‍♂️